src/backend/distributed/connection/connection_management.c

```diff
@@ -320,12 +322,58 @@ StartNodeUserDatabaseConnection(uint32 flags, const char *hostname, int32 port,
 		}
 	}
 
+
+	if (flags & NEVER_WAIT_FOR_CONNECTION)
+	{
+		/*
+		 * The caller doesn't want the connection manager to wait
+		 * until a connection slot is avaliable on the remote node.
+		 * In the end, we might fail to establish connection to the
+		 * remote node as it might not have any space in
+		 * max_connections for this connection establishment.
+		 *
+		 * Still, we keep track of the connnection counter.
+		 */
+		IncrementSharedConnectionCounter(hostname, port);
+	}
+	else if (flags & OPTIONAL_CONNECTION)
+	{
+		/*
+		 * We can afford to skip establishing an optional connection. For
+		 * non-optional connections, we first retry for some time. If we still
+		 * cannot reserve the right to establish a connection, we prefer to
+		 * error out.
+		 */
+		if (!TryToIncrementSharedConnectionCounter(hostname, port))
+		{
+			return NULL;
+		}
+	}
+	else
+	{
+		WaitOrErrorForSharedConnection(hostname, port);
+	}
+
 	/*
 	 * Either no caching desired, or no pre-established, non-claimed,
 	 * connection present. Initiate connection establishment.
 	 */
+	PG_TRY();
+	{
+		connection = StartConnectionEstablishment(&key);
+	}
+	PG_CATCH();
+	{
+		/*
+		 * Something went wrong, make sure to decement
+		 * here otherwise we'd leak the increment we have done for this
+		 * connection attempt.
+		 */
+		DecrementSharedConnectionCounter(hostname, port);
```

Comment:
Are we guaranteed that pgConn is NULL if we get here?

Reply:
My point is we're iterating over `ConnectionHashEntry *entry->connections` in `AfterXactHostConnectionHandling`, and if we throw error before adding this connection to `entry->connections`, we could lose it.

Reply:
if pgConn is not null here we'll end up decrementing again in AfterXactHostConnectionHandling. Maybe we can just call CitusPQFinish to be safe?